libraries/standard/mqtt/utest/mqtt_utest.c

```diff
@@ -2126,10 +2126,313 @@ void test_MQTT_Ping_error_path( void )
+                                                     strlen( pTopicName ),
+                                                     pTopicFilter,
+                                                     strlen( pTopicFilter ),
+                                                     &matchResult ) );
+    TEST_ASSERT_EQUAL( true, matchResult );
+
+    pTopicName = "/test/match/level";
+    pTopicFilter = "+/+/+/#";
+    TEST_ASSERT_EQUAL( MQTTSuccess, MQTT_MatchTopic( pTopicName,
+                                                     strlen( pTopicName ),
+                                                     pTopicFilter,
+                                                     strlen( pTopicFilter ),
+                                                     &matchResult ) );
+    TEST_ASSERT_EQUAL( true, matchResult );
+
+    /* Edge case where topic filter starts with '#' and topic name starts with '$'. */
+    pTopicName = "$/test/match";
+    pTopicFilter = "#";
+    TEST_ASSERT_EQUAL( MQTTSuccess, MQTT_MatchTopic( pTopicName,
+                                                     strlen( pTopicName ),
+                                                     pTopicFilter,
+                                                     strlen( pTopicFilter ),
+                                                     &matchResult ) );
+    TEST_ASSERT_EQUAL( false, matchResult );
+
+    /* Invalid topic filter where non-starting '#' is not placed after '/'.*/
+    pTopicName = "test/match/level";
+    pTopicFilter = "test/match/level#";
+    TEST_ASSERT_EQUAL( MQTTSuccess, MQTT_MatchTopic( pTopicName,
+                                                     strlen( pTopicName ),
+                                                     pTopicFilter,
+                                                     strlen( pTopicFilter ),
+                                                     &matchResult ) );
+    TEST_ASSERT_EQUAL( false, matchResult );
+
+    /* Invalid topic filters that contain '#' at a non-ending position .*/
+    pTopicName = "test/match/level1/level2";
+    pTopicFilter = "test/match/#/level2";
+    TEST_ASSERT_EQUAL( MQTTSuccess, MQTT_MatchTopic( pTopicName,
+                                                     strlen( pTopicName ),
+                                                     pTopicFilter,
+                                                     strlen( pTopicFilter ),
+                                                     &matchResult ) );
+    TEST_ASSERT_EQUAL( false, matchResult );
+
+    pTopicName = "test/match/level2";
+    pTopicFilter = "#/match/level2";
+    TEST_ASSERT_EQUAL( MQTTSuccess, MQTT_MatchTopic( pTopicName,
+                                                     strlen( pTopicName ),
+                                                     pTopicFilter,
+                                                     strlen( pTopicFilter ),
+                                                     &matchResult ) );
+    TEST_ASSERT_EQUAL( false, matchResult );
```

Comment:
Do we want to add some more unit tests for corner cases, e.g.
* topic filter: "test/match/lev#", topic name: "test/match/level1"  -> returns false
* topic filter: "test/match/#", topic name: "test/match/" -> returns true, I think
* topic filter: "test/match/lev+", topic name: "test/match/level1" -> returns false
* topic filter: "+/+", topic name: "/test" -> returns true

Just a suggestion. I'm unsure if all these specific tests need to pass, actually, since in the spec they're all "Non normative comments"

Reply:
The unit tests cover all these cases with the following:
>topic filter: "test/match/lev#", topic name: "test/match/level1"
Covered here: https://github.com/aggarw13/aws-iot-device-sdk-embedded-C/blob/mqtt/add-match-topic-api-and-unit-tests/libraries/standard/mqtt/utest/mqtt_utest.c#L2411
Though your suggestion is a variant which I can add.

>topic filter: "test/match/#", topic name: "test/match/" -> returns true, I think
https://github.com/aggarw13/aws-iot-device-sdk-embedded-C/blob/mqtt/add-match-topic-api-and-unit-tests/libraries/standard/mqtt/utest/mqtt_utest.c#L2370-L2378
(Just added)

>topic filter: "test/match/lev+", topic name: "test/match/level1" 
Covered here: https://github.com/aggarw13/aws-iot-device-sdk-embedded-C/blob/mqtt/add-match-topic-api-and-unit-tests/libraries/standard/mqtt/utest/mqtt_utest.c#L2299-L2307
Your suggestion is a variant of this, which I can add.

>topic filter: "+/+", topic name: "/test"
Covered here: https://github.com/aggarw13/aws-iot-device-sdk-embedded-C/blob/mqtt/add-match-topic-api-and-unit-tests/libraries/standard/mqtt/utest/mqtt_utest.c#L2260-L2267

Reply:
Ah I didn't see number 4. Please add 1 and 3 then, I want to make sure the tests still pass when the end of the topic name has not been reached.

Reply:
Edit, after adding the variants for cases 1 and cases 3, another bug was found in the topic matching algorithm. Have updated it. Thanks for another suggestion!!